### PR TITLE
Add Element/ShadowRoot's setHTMLUnsafe() and Document.parseHTMLUnsafe() methods

### PR DESCRIPTION
**Pull request by annevk**
#### 25a4ff15329b263c8d8550dbda463e09d2d5e41b
<pre>
Add Element/ShadowRoot&apos;s setHTMLUnsafe() and Document.parseHTMLUnsafe() methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=261143">https://bugs.webkit.org/show_bug.cgi?id=261143</a>
<a href="https://rdar.apple.com/115345128">rdar://115345128</a>

Reviewed by Ryosuke Niwa.

This change implements these methods from the HTML Standard:
<a href="https://html.spec.whatwg.org/#unsafe-html-parsing-methods">https://html.spec.whatwg.org/#unsafe-html-parsing-methods</a>

Rename FragmentScriptingPermission.h to ParserContentPolicy.h as per
the FIXME.

Rename DeclarativeShadowDOM to DeclarativeShadowRoots to match the
specification.

Abstract the implementation of Element/ShadowRoot&apos;s setInnerHTML() so
it can be reused by the new Element/ShadowRoot&apos;s setHTMLUnsafe().

Do the same for DOMParser to the benefit of Document.parseHTMLUnsafe().

Ensure that in both of those cases the new method can only go down the
HTML parser path and has declarative shadow root support enabled.

Put the new methods behind a DeclarativeShadowRootsParserAPIsEnabled
preference (enabled by default) in case something is discovered to be
amiss later on.

Synchronize web-platform-tests:
<a href="https://github.com/web-platform-tests/wpt/commit/31655b30de9bb2af8583a4d08163eae62edb4d8f">https://github.com/web-platform-tests/wpt/commit/31655b30de9bb2af8583a4d08163eae62edb4d8f</a>

Thanks to rniwa&apos;s earlier effort around the now-obsolete DOMParser API
this was all relatively straightforward. Thanks rniwa!

* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-encoding-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-encoding.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-style-attribute-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-style-attribute.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-base-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-base-pushstate-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-base-pushstate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-base.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-moretests-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-moretests.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-pushstate-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url-pushstate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe-url.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Document-parseHTMLUnsafe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Element-setHTMLUnsafe-04-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/Element-setHTMLUnsafe-04.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/parseHTMLUnsafe-iframe-base-pushstate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/parseHTMLUnsafe-iframe-base.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/parseHTMLUnsafe-iframe-pushstate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/parseHTMLUnsafe-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/parseHTMLUnsafe-iframe.js: Added.
(window.doParse):
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/parseHTMLUnsafe-url-tests.js: Added.
(assertURL):
(input.of.Object.entries):
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/resources/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/setHTMLUnsafe-CEReactions-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/setHTMLUnsafe-CEReactions.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/setHTMLUnsafe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/setHTMLUnsafe-xml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/setHTMLUnsafe-xml.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/setHTMLUnsafe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/html-unsafe-methods/w3c-import.log: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-attachment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-basic-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/DOMImplementation.cpp:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::parseMarkupUnsafe):
(WebCore::Document::parseHTMLUnsafe):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/DocumentFragment.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::replaceChildrenWithMarkup):
(WebCore::Element::setHTMLUnsafe):
(WebCore::Element::setInnerHTML):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/InnerHTML.idl:
* Source/WebCore/dom/ParserContentPolicy.h: Renamed from Source/WebCore/dom/FragmentScriptingPermission.h.
* Source/WebCore/dom/ScriptableDocumentParser.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::replaceChildrenWithMarkup):
(WebCore::ShadowRoot::setHTMLUnsafe):
(WebCore::ShadowRoot::innerHTML const):
(WebCore::ShadowRoot::setInnerHTML):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/editing/markup.cpp:
(WebCore::createFragmentForMarkup):
* Source/WebCore/editing/markup.h:
* Source/WebCore/html/HTMLTemplateElement.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::HTMLConstructionSite):
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement):
* Source/WebCore/html/parser/HTMLConstructionSite.h:
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
* Source/WebCore/xml/DOMParser.cpp:
(WebCore::DOMParser::parseFromString):
* Source/WebCore/xml/parser/XMLDocumentParser.h:
* Tools/TestWebKitAPI/Tests/WebCore/HTMLParserIdioms.cpp:

Canonical link: <a href="https://commits.webkit.org/271423@main">https://commits.webkit.org/271423@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/18a0b7267c055049880498bc157ebbb523bda66f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28239 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6882 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30770 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25733 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28736 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4262 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24305 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4924 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31459 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/24436 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25860 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25738 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31359 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/27349 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5024 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3241 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29120 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6583 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/25100 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/34859 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6789 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5461 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7543 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5542 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/dom/Document.idl [hunk 1]
```diff
@@ -41,6 +41,8 @@ typedef (HTMLScriptElement or SVGScriptElement) HTMLOrSVGScriptElement;
 ] interface Document : Node {
     [CallWith=CurrentDocument] constructor();
 
+    [CallWith=CurrentScriptExecutionContext, EnabledBySetting=DeclarativeShadowRootsParserAPIsEnabled] static Document parseHTMLUnsafe(DOMString html);
```
Comment (annevk) [inline Source/WebCore/dom/Document.idl]: After initial review I'll clean this up to use `CurrentDocument` which also avoids a downcast in `Document`.
```suggestion
    [CallWith=CurrentDocument, EnabledBySetting=DeclarativeShadowRootsParserAPIsEnabled] static Document parseHTMLUnsafe(DOMString html);
```

**Comment by annevk** [comment]
This is implementing https://html.spec.whatwg.org/#unsafe-html-parsing-methods btw. I'll add that to the final commit message.

File: Source/WebCore/dom/Document.cpp [hunk 2]
```diff
@@ -906,6 +907,33 @@ void Document::commonTeardown()
...
+            html->appendChild(body);
+        }
+    }
+    if (!usedFastPath)
+        setContent(markup);
+}
+
+Ref<Document> Document::parseHTMLUnsafe(ScriptExecutionContext& context, const String& html)
+{
+    Ref document = DOMImplementation::createDocument("text/html"_s, nullptr, downcast<Document>(context).protectedSettings(), URL { });
```
Comment (rniwa) [inline Source/WebCore/dom/Document.cpp]: This is a very roundabout way of calling `HTMLDocument::create(frame, settings, url, documentIdentifier)`. We should just call that function directly.

File: Source/WebCore/dom/Document.cpp [hunk 3]
```diff
@@ -906,6 +907,33 @@ void Document::commonTeardown()
...
+    bool usedFastPath = false;
+    if (this->contentType() == "text/html"_s) {
+        auto body = HTMLBodyElement::create(*this);
+        usedFastPath = tryFastParsingHTMLFragment(StringView { markup }.substring(markup.find(isNotASCIIWhitespace<UChar>)), *this, body, body, policy);
+        if (LIKELY(usedFastPath)) {
+            auto html = HTMLHtmlElement::create(*this);
+            appendChild(html);
+            auto head = HTMLHeadElement::create(*this);
+            html->appendChild(head);
+            html->appendChild(body);
```
Comment (rniwa) [inline Source/WebCore/dom/Document.cpp:922]: This is inefficient. We'd do O(n) traversal to update the node flags, etc... we should add body into html beforehand.

File: Source/WebCore/dom/ParserContentPolicy.h [hunk 4]
```diff
@@ -35,10 +33,11 @@ enum class ParserContentPolicy : uint8_t {
     AllowScriptingContent = 1 << 0,
     AllowPluginContent = 1 << 1,
     DoNotMarkAlreadyStarted = 1 << 2,
-    AllowDeclarativeShadowDOM = 1 << 3,
+    AllowDeclarativeShadowRoots = 1 << 3,
+    ParseAsHTMLOnly = 1 << 4
```
Comment (rniwa) [inline Source/WebCore/dom/ParserContentPolicy.h]: I'd call this AlwaysParseAsHTML or something.

**Review by rniwa — APPROVED** [review]

**Pull request merged**